File: web/src/features/availability.ts
/**
 * The availability control — how the player tells the app what they can field.
 *
 * Exclusion mode is presented first and needs no setup: name a card you don't have,
 * or tick a whole class ("no Epics"), and every card list and deck readout updates.
 * Collection mode is the precise-but-expensive option, offered second.
 */

import type {
  AvailabilityMode,
  AvailabilityProfile,
  CardFacets,
  ExcludedCard,
} from "../api/types";
import {
  forgetCollection,
  setAvailabilityMode,
  setStrict,
  toggleOwnedRule,
  toggleRule,
  unexcludeCard,
} from "../state/actions";
import { store } from "../state/store";
import { h, replace } from "../ui/dom";
import { collectionSummary } from "./collectionSummary";

const MODE_LABELS: Record<AvailabilityMode, { title: string; hint: string }> = {
  exclusion: {
    title: "Cards I don't have",
    hint: "Name what you're missing. Everything else is fair game — including new sets.",
  },
  collection: {
    title: "My collection",
    // Was "Precise, but you have to record what you own first" -- true when the only
    // writer was the wizard's opt-in write-back, and a dead end: the mode told the
    // player to record something and gave them nowhere to do it.
    hint: "Use recorded quantities, or count a complete rarity or set as available.",
  },
  open: {
    title: "Everything",
    hint: "Build with the whole card pool.",
  },
};

const MODE_ORDER: AvailabilityMode[] = ["exclusion", "collection", "open"];
let lastProfile: AvailabilityProfile | null = null;
let lastFacets: CardFacets | null = null;
let lastRoot: HTMLElement | null = null;

function modeButton(mode: AvailabilityMode, active: boolean): HTMLElement {
  const label = MODE_LABELS[mode];
  return h(
    "button",
    {
      class: `mode-btn${active ? " is-active" : ""}`,
      type: "button",
      data: { availabilityKey: `mode:${mode}` },
      title: label.hint,
      aria: { pressed: String(active) },
      on: { click: () => void setAvailabilityMode(mode) },
    },
    label.title,
  );
}

function excludedChip(entry: ExcludedCard): HTMLElement {
  return h(
    "span",
    { class: "chip" },
    entry.name,
    h(
      "button",
      {
        class: "chip-x",
        type: "button",
        title: "I do have this after all",
        aria: { label: `Remove ${entry.name}` },
        on: { click: () => void unexcludeCard(entry.cardId) },
      },
      "×",
    ),
  );
}

/** One-click rules, built from the values actually present in the bundle. */
function quickRules(profile: AvailabilityProfile, facets: CardFacets | null): HTMLElement {
  const active = (kind: string, value: string) =>
    profile.rules.some((r) => r.kind === kind && r.value === value);

  const options: { kind: string; value: string; label: string }[] = [
    ...(facets?.rarities ?? [])
      .filter((r) => r === "Epic" || r === "Rare" || r === "Showcase")
      .map((r) => ({ kind: "rarity", value: r, label: `No ${r}s` })),
    { kind: "promo_only", value: "", label: "No promo-only cards" },
    ...(facets?.setCodes ?? []).map((s) => ({
      kind: "set",
      value: s,
      label: `No ${s}`,
    })),
  ];

  return h(
    "div",
    { class: "quick-rules" },
    ...options.map((option) =>
      h(
        "button",
        {
          class: `pill${active(option.kind, option.value) ? " is-on" : ""}`,
          type: "button",
          aria: { pressed: String(active(option.kind, option.value)) },
          data: { availabilityKey: `excluded:${option.kind}:${option.value}` },
          on: { click: () => void toggleRule(option.kind, option.value) },
        },
        option.label,
      ),
    ),
  );
}

/**
 * One-click rules for what the player *does* have.
 *
 * The mirror of `quickRules`, and the half that was missing. Until this existed,
 * "My collection" told the player to record what they own and gave them nowhere to do
 * it: the only writer in the app was the wizard's opt-in write-back, so a collection
 * could only ever contain cards some session happened to ask about. Naming what you
 * lack is the right shape for somebody who owns nearly everything; a casual player
 * would have to list thousands of cards to say something true about a few hundred.
 */
function ownedRules(profile: AvailabilityProfile, facets: CardFacets | null): HTMLElement {
  const active = (kind: string, value: string) =>
    profile.ownedRules.some((r) => r.kind === kind && r.value === value);

  const options: { kind: string; value: string; label: string }[] = [
    // Rarity first: it is the axis a collection actually thins out along.
    ...(facets?.rarities ?? [])
      .filter((r) => r === "Common" || r === "Uncommon" || r === "Rare")
      .map((r) => ({ kind: "rarity", value: r, label: `All ${r}s` })),
    ...(facets?.setCodes ?? []).map((code) => ({
      kind: "set",
      value: code,
      label: `All of ${code}`,
    })),
  ];

  return h(
    "div",
    { class: "quick-rules" },
    ...options.map((option) =>
      h(
        "button",
        {
          class: `pill${active(option.kind, option.value) ? " is-on" : ""}`,
          type: "button",
          aria: { pressed: String(active(option.kind, option.value)) },
          data: { availabilityKey: `owned:${option.kind}:${option.value}` },
          on: { click: () => void toggleOwnedRule(option.kind, option.value) },
        },
        option.label,
      ),
    ),
  );
}


export function renderAvailability(root: HTMLElement): void {
  const { availability: profile, facets } = store.state;
  if (root === lastRoot && profile === lastProfile && facets === lastFacets) return;
  lastRoot = root;
  lastProfile = profile;
  lastFacets = facets;
  const activeKey = (document.activeElement as HTMLElement | null)?.dataset.availabilityKey;
  if (!profile) {
    replace(root, h("p", { class: "muted" }, "Loading…"));
    return;
  }

  const summary = collectionSummary(profile);
  const body: HTMLElement[] = [
    h("p", { class: "availability-summary" }, h("strong", {}, summary.label), summary.detail),
    h(
      "div",
      { class: "mode-row" },
      h("div", { class: "mode-group", role: "group" },
        ...MODE_ORDER.map((mode) => modeButton(mode, profile.mode === mode))),
      h(
        "label",
        { class: "strict-toggle", title: "Hide unavailable cards entirely instead of ranking them lower" },
        h("input", {
          type: "checkbox",
          checked: profile.strict,
          data: { availabilityKey: "strict" },
          on: { change: (e) => void setStrict((e.target as HTMLInputElement).checked) },
        }),
        " Only what I can build now",
      ),
    ),
    h("p", { class: "mode-hint" }, MODE_LABELS[profile.mode].hint),
  ];

  if (profile.mode === "exclusion") {
    body.push(
      h(
        "div",
        { class: "exclusions" },
        profile.excludedCards.length > 0
          ? h(
              "div",
              { class: "chips" },
              h("span", { class: "chips-label" }, "Missing:"),
              ...profile.excludedCards.map(excludedChip),
            )
          : h(
              "p",
              { class: "muted small" },
              "Click “I don't have this” on any card to start.",
            ),
        quickRules(profile, facets),
      ),
    );
  } else if (profile.mode === "collection") {
    const counted =
      profile.ownedCardCount > 0
        ? `${profile.ownedCardCount} distinct cards recorded`
        : "";
    const declared = profile.ownedRules.length
      ? profile.ownedRules.map((r) => r.description).join(", ")
      : "";
    body.push(
      h(
        "div",
        { class: "exclusions" },
        h(
          "p",
          { class: "muted small" },
          counted && declared
            ? `You have ${declared}, plus ${counted}.`
            : declared
              ? `You have ${declared}.`
              : counted
                ? `${counted}.`
                : "Use the deck finder to record quantities as you go, or add a collection shortcut below.",
        ),
        ownedRules(profile, facets),
        h("p", { class: "muted small" }, "Shortcuts assume full quantities of every matching card, including future sets. They are combined: selecting a rarity and a set includes both groups."),
      ),
    );
  }

  // The way back out. The wizard offers to write what a session learned into the
  // collection, which is much the fastest way to record one — and a one-way door into
  // that is not a fair trade for the convenience. Always offered, not only in collection
  // mode, because a session's answers outlive the mode that produced them.
  body.push(
    h(
      "div",
      { class: "forget-collection" },
      h(
        "button",
        {
          class: "quiet-button",
          type: "button",
          title: "Delete the recorded collection and every Smart Decks session",
          on: { click: () => {
            if (window.confirm("Reset your recorded collection and delete all deck-finder sessions? Saved decks are kept. This cannot be undone.")) void forgetCollection();
          } },
        },
        "Reset collection & finder sessions…",
      ),
      h(
        "span",
        { class: "muted small" },
        "Removes the recorded collection and every saved wizard session.",
      ),
    ),
  );

  body.push(h("p", { class: "availability-storage-note" }, "Saved decks and collection records belong to this browser. Clearing site cookies loses access to them. Export decks or copy their links to keep a backup."));

  replace(root, ...body);
  if (activeKey) Array.from(root.querySelectorAll<HTMLElement>("[data-availability-key]"))
    .find((element) => element.dataset.availabilityKey === activeKey)?.focus({ preventScroll: true });
}

File: server/riftbound/api/routes/availability.py
"""The availability profile: the player's answer to "what can you actually field?"

The exclusion endpoints are the low-friction path. ``POST /api/availability/exclude``
takes a card the player says they don't have and switches the profile into exclusion
mode on the spot, so the first thing a new player does can be "I don't have Seal of
Discord" -- no collection, no setup screen.
"""

from __future__ import annotations

from dataclasses import replace

from fastapi import APIRouter, Depends, HTTPException

from ...domain.availability import (
    MODE_COLLECTION,
    MODE_EXCLUSION,
    RULE_KINDS,
    AvailabilityProfile,
    ExclusionRule,
    OwnedRule,
)
from ...services import Services, get_services
from ..identity import Identity, current_identity
from ..schemas import AvailabilityUpdate, AvailabilityView, ForgetResult
from ..views import availability_view

router = APIRouter(prefix="/api/availability", tags=["availability"])


def _load(services: Services, user_id: str) -> AvailabilityProfile:
    return services.availability.load(user_id=user_id)


@router.get("", response_model=AvailabilityView)
def get_availability(
    services: Services = Depends(get_services),
    identity: Identity = Depends(current_identity),
) -> AvailabilityView:
    return availability_view(_load(services, identity.user_id), services.catalog)


@router.delete("/collection", response_model=ForgetResult)
def forget_collection(
    services: Services = Depends(get_services),
    identity: Identity = Depends(current_identity),
) -> ForgetResult:
    """Erase everything recorded about what this user owns.

    Both halves, deliberately. The collection rows are the obvious ones, but a wizard
    session's answers say what somebody owns just as plainly -- three rounds pin down
    roughly 75 cards -- so erasing one and leaving the other would be a privacy control
    that only looks like one.

    The wizard offers to write what a session learned into the collection, which is by
    some distance the fastest way to record one. A one-way door into that is not a fair
    trade for the convenience, so this is the way back out.

    Collection mode is switched to open on the way, because a collection-mode profile
    standing on an empty collection tells the player every card in the game is missing --
    which is true, and useless. Open is the honest posture once they have told us
    nothing: assume they can get anything, rather than that they have nothing.
    """
    rows = services.collections.clear(user_id=identity.user_id)
    sessions = services.smart_decks.clear_all(user_id=identity.user_id)

    profile = _load(services, identity.user_id)
    profile = replace(
        profile,
        mode="open" if profile.mode == MODE_COLLECTION else profile.mode,
        owned={},
        owned_rules=(),
    )
    services.availability.save(profile, user_id=identity.user_id)

    return ForgetResult(
        collection_rows=rows,
        sessions=sessions,
        availability=availability_view(profile, services.catalog),
    )


@router.put("", response_model=AvailabilityView)
def set_availability(
    update: AvailabilityUpdate,
    services: Services = Depends(get_services),
    identity: Identity = Depends(current_identity),
) -> AvailabilityView:
    try:
        mode = update.validated_mode()
        rules = update.validated_rules()
        owned_rules = update.validated_owned_rules()
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    unknown = [cid for cid in update.excluded_card_ids if services.catalog.get(cid) is None]
    if unknown:
        raise HTTPException(
            status_code=400,
            detail=f"Unknown card id(s): {', '.join(sorted(unknown)[:5])}",
        )
    # Mode chooses which settings apply. It must not erase the inactive settings
    # the client carries across a mode switch.
    profile = AvailabilityProfile(
        mode=mode,
        owned=services.collections.owned_by_card(user_id=identity.user_id, include_zero=True),
        excluded_cards=frozenset(update.excluded_card_ids),
        exclusion_rules=tuple(ExclusionRule(kind=r.kind, value=r.value) for r in rules),
        owned_rules=tuple(OwnedRule(kind=r.kind, value=r.value) for r in owned_rules),
        strict=update.strict,
        penalty=update.penalty,
    )

    services.availability.save(profile, user_id=identity.user_id)
    return availability_view(profile, services.catalog)


@router.post("/exclude/{card_id}", response_model=AvailabilityView)
def exclude_card(
    card_id: str,
    services: Services = Depends(get_services),
    identity: Identity = Depends(current_identity),
) -> AvailabilityView:
    """"I don't have this one." Switches into exclusion mode if needed."""
    card = services.catalog.get(card_id)
    if card is None:
        raise HTTPException(status_code=404, detail=f"No card {card_id!r}.")

    current = _load(services, identity.user_id)
    profile = replace(
        current,
        mode=MODE_EXCLUSION,
        excluded_cards=current.excluded_cards | {card.card_id},
        strict=current.strict if current.mode == MODE_EXCLUSION else False,
        penalty=current.penalty,
    )
    services.availability.save(profile, user_id=identity.user_id)
    return availability_view(profile, services.catalog)


@router.delete("/exclude/{card_id}", response_model=AvailabilityView)
def unexclude_card(
    card_id: str,
    services: Services = Depends(get_services),
    identity: Identity = Depends(current_identity),
) -> AvailabilityView:
    current = _load(services, identity.user_id)
    profile = replace(
        current,
        mode=MODE_EXCLUSION,
        excluded_cards=current.excluded_cards - {card_id.strip().lower()},
        strict=current.strict,
        penalty=current.penalty,
    )
    services.availability.save(profile, user_id=identity.user_id)
    return availability_view(profile, services.catalog)


@router.get("/rule-kinds")
def rule_kinds(services: Services = Depends(get_services)) -> dict:
    """Available one-click exclusion rules, with the values present in this bundle."""
    catalog = services.catalog
    return {
        "kinds": list(RULE_KINDS),
        "values": {
            "rarity": sorted({c.rarity for c in catalog if c.rarity}),
            "set": sorted({s for c in catalog for s in c.set_codes}),
            "super_type": sorted({c.super_type for c in catalog if c.super_type}),
            "promo_only": [],
        },
    }
